aeppl/scan.py

```diff
@@ -0,0 +1,326 @@
+        var_idxs.append(_inner_out_info.index)
+        inner_out_info.append(_inner_out_info)
+
+    # The old inner-output variable becomes the a new inner-input
+    inner_in_var = old_inner_out_var.clone()
+
+    # We need to clone any existing inner-output variables in the `ScanArgs`
+    # object that we're mutating and replace references to `old_inner_out_var`
+    # with `inner_in_var`.  If we don't, then any other inner-outputs that
+    # reference the inner-output that we're replacing will be inconsistent.
+    # Instead, we want those other inner-outputs to reference the new
+    # inner-input replacement variable.
+    for io_var in list(output_scan_args.inner_outputs):
+        io_var_info = output_scan_args.find_among_fields(
+            io_var, field_filter=lambda x: x.startswith("inner_out")
+        )
+        io_sub_list = getattr(output_scan_args, io_var_info.name)
+
+        (new_io_var,) = clone_replace(
+            [io_var], replace={old_inner_out_var: inner_in_var}
+        )
+
+        io_sub_list[io_var_info.index] = new_io_var
+
+    # If we're replacing a [m|s]it-sot, then we need to add a new nit-sot
+    add_nit_sot = False
+    inner_in_seqs = [inner_in_var]
+    if inner_out_info.name.endswith("mit_sot"):
+        inner_in_seqs = input_scan_args.inner_in_mit_sot[var_idxs] + inner_in_seqs
+        if replacing:
+            output_scan_args.inner_in_mit_sot.pop(var_idxs)
+            output_scan_args.outer_in_mit_sot.pop(var_idxs)
+        add_nit_sot = True
+    elif inner_out_info.name.endswith("sit_sot"):
+        inner_in_seqs = [input_scan_args.inner_in_sit_sot[var_idxs]] + inner_in_seqs
+        if replacing:
+            output_scan_args.inner_in_sit_sot.pop(var_idxs)
+            output_scan_args.outer_in_sit_sot.pop(var_idxs)
+        add_nit_sot = True
+
+    taps = [0]
+    if inner_out_info.name.endswith("mit_sot"):
+        taps = input_scan_args.mit_sot_in_slices[var_idxs] + taps
+        if replacing:
+            output_scan_args.mit_sot_in_slices.pop(var_idxs)
+    elif inner_out_info.name.endswith("sit_sot"):
+        taps = [-1] + taps
+
+    taps, inner_in_seqs = zip(*sorted(zip(taps, inner_in_seqs), key=lambda x: x[0]))
+
+    inner_in_seqs = list(reversed(inner_in_seqs))
+    output_scan_args.inner_in_seqs += inner_in_seqs
+
+    slice_seqs = zip(-np.asarray(taps), [n if n < 0 else None for n in reversed(taps)])
+
+    var_slices = [new_outer_input_vars[b:e] for b, e in slice_seqs]
+    n_steps = at.min([at.shape(n)[0] for n in var_slices])
+
+    if output_scan_args.n_steps is None or replacing:
+        output_scan_args.n_steps = n_steps
+
+    output_scan_args.outer_in_seqs += [v[:n_steps] for v in var_slices]
+
+    if not replacing or add_nit_sot:
+        output_scan_args.outer_in_nit_sot += [n_steps]
+
+    if inner_out_fn:
+        output_scan_args.inner_out_nit_sot += [
+            inner_out_fn(
+                input_scan_args, old_inner_out_var, inner_in_var, output_scan_args
```

Comment:
This is where we should be passing the processed value of corresponding RV node using `new_outer_input_vars`. Issue is how do we make sure that each iteration of `Scan` in its `n_step`s of iterations receives the correct value from this array of value.  (i.e. we'll probably need to modify the above `taps` logic for that which will correctly pass the values for each output node as a `seq` type input.)

Reply:
Yeah, there are some design questions/issues with `convert_outer_out_to_in` and multiple value variables.  

I believe `convert_outer_out_to_in` I designed to work with one value variable at a time via the `ScanArgs` (i.e. one would mutate the `ScanArgs` in-place for each random variable).

It should repurpose the original `Scan` so that each new value variable is a new outer-input and&mdash;thus&mdash;has a corresponding inner-input "slice"/"scalar".  All the `taps` logic should be handled by that existing logic, as long as the value variable replacements we use are correct.

These replacements probably need to be the un-`*Subtensor*`ed tensors, and not the ones returned by `aesara.scan`, which cut off the initial values via `*Subtensor*` slices.